### ISSUE TITLE
Consider options for decorator nodes_or_number

### ISSUE THREAD

**Issue by dschult** [networkx/networkx#5612]
Recent Issue #5582 and PR #5599 have given me a chance to look at the decorator `nodes_or_number` more closely.  I can envision different improvements that go in different directions.

- many functions use this decorator and then quickly call `empty_graph` with results. But `empty_graph` is also decorated, so the same computations are used again. But removing the decorator and just feeding the input to `empty_graph` means that you must check whether the input `n` was an integer or a sequence of nodes. Such processing is fairly easy:
```python
G = empty_graph(n, create_using)
nodes = G.nodes if n == len(G) else n
... then make e.g. a path with edges built from the sequence `nodes`
```

This is not easier to read than when using the decorator:
```python
n, nodes = n
G = empty_graph(nodes, create_using)
... then make e.g. a path with edges built from the sequence `nodes`
```

Perhaps the best solution is to update nodes_or_number to have an optional argument `original_input_too=False`. This would remove the need to the line `n, nodes = n` at the top of the functions that have this decorator. For the 2 or 3 functions that need both outputs one can set the argument so you can get the original input as well as the processed input.

Another solution would be to remove the `nodes_or_number` decorator altogether. This would add some code to `empty_graph`, `lollipop_graph` and `complete_bipartite_graph`, but not much code. The danger here is a decentralization of input handling so over time we could easily get different behaviors for different functions that accept an input as either a number of nodes or a sequence of nodes.

**Comment by rossbar** [comment]
> Another solution would be to remove the nodes_or_number decorator altogether. This would add some code to empty_graph, lollipop_graph and complete_bipartite_graph, but not much code. The danger here is a decentralization of input handling so over time we could easily get different behaviors for different functions that accept an input as either a number of nodes or a sequence of nodes.

If it's really only 3 (or a similarly small number) of functions that would really be affected, then I think it's worth at least considering removing the `nodes_or_number` decorator. In the long run it might be a good thing to discourage new functions from adopting the API that allows *either* an integer specifying the number of nodes or a container of nodes and forcing authors to choose one or the other, though there may be cases that warrant the flexibility.

**Comment by dschult** [comment]
Removing `nodes_or_number` won't discourage specifying either nodes or a number as an input. It would just relocate the code where that option is provided from the decorator to the `empty_graph` function.  Removing the option to provide either a number of nodes or a collection of nodes would be a different discussion.  This question is more about where we should store the code that provides this option.

**Comment by rossbar** [comment]
> Removing nodes_or_number won't discourage specifying either nodes or a number as an input. It would just relocate the code where that option is provided from the decorator to the empty_graph function.

Good point - I was thinking about nodes or number of nodes as input for functions more generally, but in practice this is really only relevant for the graph generator functions.

> Removing the option to provide either a number of nodes or a collection of nodes would be a different discussion. This question is more about where we should store the code that provides this option.

Agreed - the focus should be on the latter - please ignore any of my comments that tend towards expanded scope!

**Comment by boothby** [comment]
I've used the `nodes_or_number` in a downstream package, because I find it convenient: https://github.com/dwavesystems/minorminer/blob/cc5bb76341cc71bc8730be465dd37b20a494a420/minorminer/utils/zephyr.py#L72

It would be easy to manage if the decorator was removed, but since the function is in a non-underscore module, a deprecation would be proper.

**Comment by dschult** [comment]
With regard to removing `nodes_or_number` from functions that currently call `empty_graph` immediately.
It may not be as easy as I thought to remove `nodes_or_number`. It changes behavior when duplicates appear in the incoming list of nodes.  We currently return incorrect graphs in this case!!

For example, `complete_graph` currently doesn't return the complete graph K_n (despite saying that it does) when the input container of nodes has duplicates in it. Self-loops are possible if nodes are duplicates in the container of nodes!
```python
G=nx.complete_graph("abcb")
print(G.edges)
```
produces output: `EdgeView([('a', 'b'), ('a', 'c'), ('b', 'c'), ('b', 'b')])`

If we remove the decorator from these functions they will correctly produce a complete graph on the set of nodes indicated by the input node container **ignoring duplicates**.  This is actually better in the sense that the result of e.g. `complete_graph` will be a complete graph!  But it is **not** what we currently provide. 

Do we have to deprecate to protect those who incorrectly provide a list of nodes that contains duplicates? Even though they probably are getting a silently incorrect graph?   I'd slightly prefer to just "fix" the current incorrect behavior than to deprecate and wait until 3.2 to make the fix. But I'm willing to deprecate if that is deemed the better path.
